aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -1664,101 +1666,119 @@ Tensor _matmul_impl(
     return has_out ? at::mv_out(out, tensor2.t(), tensor1) : tensor2.t().mv(tensor1);
   } else if (dim_tensor1 == 2 && dim_tensor2 == 2) {
     return has_out ? at::mm_out(out, tensor1, tensor2) : tensor1.mm(tensor2);
-  } else if (should_fold_into_mm(tensor1, tensor2)) {
-    // optimization: use mm instead of bmm by folding tensor1's batch into
-    // its leading matrix dimension.
-
-    Tensor t2 = dim_tensor2 == 1 ? tensor2.unsqueeze(-1) : tensor2;
-    auto size1 = tensor1.sizes();
-    auto size2 = t2.sizes();
-    DimVector output_size(size1.begin(), size1.end() - 1);
-    if (dim_tensor2 > 1) {
-      output_size.push_back(size2[dim_tensor2 - 1]);
-    }
-
-    // fold the batch into the first dimension
-    // Why not tensor1.view(-1, size1[size1.size() -1])?
+  } else if (should_fold(tensor1, dim_tensor2) || should_fold(tensor2, dim_tensor1)) {
```

Comment:
so I'd suggest we don't try to handle 2d-3d case now, given all the extra copies that it incurs it's better to avoid this, and it'll also make this PR simpler.

Reply:
All this is fixed in the next PR of the stack, so I suggest we leave it as it is, as this PR is not changing the inner dispatch of the operation, it just tries to do the same thing as the code before but faster.